js/source/source.js

```diff
@@ -3,70 +3,32 @@
 var util = require('../util/util');
 var ajax = require('../util/ajax');
 var browser = require('../util/browser');
-var TilePyramid = require('./tile_pyramid');
 var normalizeURL = require('../util/mapbox').normalizeSourceURL;
 var TileCoord = require('./tile_coord');
 
-exports._loadTileJSON = function(options) {
+exports._loadTileJSON = function(options, callback) {
     var loaded = function(err, tileJSON) {
         if (err) {
-            this.fire('error', {error: err});
-            return;
+            return callback(err);
         }
 
-        util.extend(this, util.pick(tileJSON,
-            ['tiles', 'minzoom', 'maxzoom', 'attribution']));
+        var result = util.pick(tileJSON, ['tiles', 'minzoom', 'maxzoom', 'attribution']);
 
         if (tileJSON.vector_layers) {
-            this.vectorLayers = tileJSON.vector_layers;
-            this.vectorLayerIds = this.vectorLayers.map(function(layer) { return layer.id; });
+            result.vectorLayers = tileJSON.vector_layers;
+            result.vectorLayerIds = result.vectorLayers.map(function(layer) { return layer.id; });
         }
 
-        this._pyramid = new TilePyramid({
-            tileSize: this.tileSize,
-            minzoom: this.minzoom,
-            maxzoom: this.maxzoom,
-            roundZoom: this.roundZoom,
-            reparseOverscaled: this.reparseOverscaled,
-            load: this._loadTile.bind(this),
-            abort: this._abortTile.bind(this),
-            unload: this._unloadTile.bind(this),
-            add: this._addTile.bind(this),
-            remove: this._removeTile.bind(this),
-            redoPlacement: this._redoTilePlacement ? this._redoTilePlacement.bind(this) : undefined
-        });
-
-        this.fire('load');
-    }.bind(this);
+        callback(null, result);
+    };
 
     if (options.url) {
         ajax.getJSON(normalizeURL(options.url), loaded);
     } else {
-        browser.frame(loaded.bind(this, null, options));
-    }
-};
-
-exports.redoPlacement = function() {
-    if (!this._pyramid) {
-        return;
-    }
-
-    var ids = this._pyramid.orderedIDs();
-    for (var i = 0; i < ids.length; i++) {
-        var tile = this._pyramid.getTile(ids[i]);
-        this._redoTilePlacement(tile);
+        browser.frame(loaded.bind(null, null, options));
     }
 };
```

Comment:
Do you think it'd make sense to put `loadTileJSON`, `queryRenderedFeatures`, and `querySourceFeatures` into standalone files? They don't gain much by being coupled together with `Source.create` and the source type directory.